### PR TITLE
[export] Add support for symbool to make it usable for torch.cond

### PR DESCRIPTION
**Pull request by henrylhtsang**
# Why?

I want the following code to work.

minimal repro:
```
class M(torch.nn.Module):
    def forward(self, dilate_flag):
        return dilate_flag.item()

input1 = (torch.tensor([1], dtype=torch.bool, device="cuda"),)
model = M().cuda()

ep = torch.export.export(model, input1, strict=True)
path = torch._inductor.aot_compile(ep.module(), input1)
aot_model = torch._export.aot_load(path, device="cuda")
actual_output = aot_model(*input1)
```

error: AssertionError: Encountered an unsupported object of type <class 'torch.SymBool'> while writing the metadata for exported program

second error will be handled by https://github.com/pytorch/pytorch/pull/138760

# Motivation

I could technically bypass it with a torch.int tensor. However, it doesn't work with torch.cond. I want the following to work. It would also require https://github.com/pytorch/pytorch/pull/138760 for aot compile to work.

```
class M(torch.nn.Module):
    def __init__(self) -> None:
        super().__init__()
        self.dilate_flag = 0

    def forward(self, dilate_flag):
        self.dilate_flag = dilate_flag.item()

        def true_fn(dilate_flag):
            return dilate_flag.clone()

        def false_fn(dilate_flag):
            return dilate_flag.clone()

        torch.cond(
            self.dilate_flag,
            true_fn,
            false_fn,
            (dilate_flag,),
        )
        return self.dilate_flag

input1 = (torch.tensor([1], dtype=torch.bool, device="cuda"),)
input2 = (torch.tensor([0], dtype=torch.bool, device="cuda"),)
inputs = (input1, input2)
model = M().cuda()

for input in inputs:
    expected_output = model(*input)

    ep = torch.export.export(model, input, strict=False)
    path = torch._inductor.aot_compile(ep.module(), input)
    aot_model = torch._export.aot_load(path, device="cuda")
    actual_output = aot_model(*input)

    assert (
        expected_output == actual_output
    ), f"henry they are not equal {expected_output} != {actual_output}"
```

Differential Revision: D64867504

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov

### CODE REVIEW

File: torch/export/graph_signature.py [hunk 1]
```diff
@@ -472,6 +472,8 @@ def _make_argument_spec(node, token_names) -> ArgumentSpec:
         return TensorArgument(name=node.name)
     elif isinstance(val, SymInt):
         return SymIntArgument(name=node.name)
+    elif isinstance(val, SymBool):
+        return SymIntArgument(name=node.name)
```
[the remark this answers is not in the document]
Comment (ydwu4) [reply torch/export/graph_signature.py:476]: cc @angelayi for help on the implementation

**Comment by henrylhtsang** [comment]
@ydwu4 Added Symbool argument, added tests in export and aot inductor.

**Comment by henrylhtsang** [comment]
@ydwu4 I am getting the following rst error. Any idea?

```

Testing of coverage in the sources finished, look at the results in build/coverage/python.txt.
/opt/conda/envs/py_3.9/lib/python3.9/tempfile.py:830: ResourceWarning: Implicitly cleaning up <TemporaryDirectory '/tmp/tmpjw7iy0ci'>
  _warnings.warn(warn_message, ResourceWarning)
++ wc -l build/coverage/python.txt
++ cut -f1 '-d '
+ lines=7
+ undocumented=5
+ '[' 5 -lt 0 ']'
+ '[' 5 -gt 0 ']'
+ echo undocumented objects found:
undocumented objects found:
+ cat build/coverage/python.txt
Undocumented Python objects
===========================
torch.export.graph_signature
----------------------------
Classes:
 * SymBoolArgument

+ echo 'Make sure you'\''ve updated relevant .rsts in docs/source!'
Make sure you've updated relevant .rsts in docs/source!
+ echo 'You can reproduce locally by running '\''cd docs && make coverage && cat build/coverage/python.txt'\'''
You can reproduce locally by running 'cd docs && make coverage && cat build/coverage/python.txt'
+ exit 1
```

File: test/inductor/test_aot_inductor_arrayref.py [hunk 2]
```diff
@@ -173,6 +173,10 @@ def fail_minimal_arrayref_interface(is_skip=False):
     ),
     "test_size_from_multi_output": fail_stack_allocation(is_skip=True),
     "test_torchvision_transforms_functional_tensor_resize": fail_minimal_arrayref_interface(),
+    # TODO: AttributeError: 'ShapeAsConstantBuffer' object has no attribute 'dtype'
```
Comment (ydwu4) [inline test/inductor/test_aot_inductor_arrayref.py:176]: Can you paste the full error stack trace?  cc @desertfire for review on this change.

File: torch/_export/verifier.py [hunk 3]
```diff
@@ -307,7 +308,7 @@ def _verify_exported_program_signature(exported_program) -> None:
         )
 
     for input_spec, node in zip(gs.input_specs, input_node_names):
-        if isinstance(input_spec.arg, (TensorArgument, SymIntArgument)):
+        if isinstance(input_spec.arg, (TensorArgument, SymIntArgument, SymBoolArgument)):
```
Comment (ydwu4) [inline torch/_export/verifier.py:311]: When we have SymBoolArgument inputs? Can add a test for it?

**Comment by ydwu4** [comment]
> the following rst error

Appears to be some doc issue.  Not familiar with the doc issues. Probably can follow https://github.com/pytorch/pytorch/blob/main/docs/source/export.rst?plain=1#L896 . But not sure why this is not triggered for SymIntArgument. Might worth dig deeper.

File: torch/_export/verifier.py (hunk 3 above)
Comment (henrylhtsang) [reply torch/_export/verifier.py:311]: @ydwu4 I am not sure, I was just following the pattern of SymIntArgument.

Any SymIntArgument test that I can take a look?

File: test/inductor/test_aot_inductor_arrayref.py (hunk 2 above)
Comment (henrylhtsang) [reply test/inductor/test_aot_inductor_arrayref.py:176]: stack trace: https://gist.github.com/henrylhtsang/a6ef1a85c56c9fafef1cbc188ca82e67

```
/torch/_inductor/codegen/cpp_wrapper_cpu_array_ref.py", line 134, in <genexpr>
    f"ArrayRefTensor<{DTYPE_TO_CPP[x.get_dtype()]}>"
  /torch/_inductor/ir.py", line 393, in get_dtype
    return self.dtype
AttributeError: 'ShapeAsConstantBuffer' object has no attribute 'dtype'
```

**Review by ydwu4 — APPROVED** [review]
Add another test case for test_lifted_constants of boolean?

**Comment by henrylhtsang** [comment]
> Add another test case for test_lifted_constants of boolean?

my bad, added now

**Pull request closed without merging**